go/vt/vttablet/tabletmanager/tm_init.go

```diff
@@ -384,8 +386,17 @@ func (tm *TabletManager) createKeyspaceShard(ctx context.Context) (*topo.ShardIn
 	srvKeyspace, err := tm.TopoServer.GetSrvKeyspace(ctx, tm.tabletAlias.Cell, tablet.Keyspace)
 	switch {
 	case err == nil:
-		tm.tmState.RefreshFromTopoInfo(ctx, nil, srvKeyspace)
+		if proto.Equal(srvKeyspace, &topodatapb.SrvKeyspace{}) {
```

Comment:
Better to check if the current keyspace exists. The SrvKeyspace could be non-empty, but not have the current keyspace.

Reply:
We are already checking that (line 386). There is a file per-keyspace and we will get a topo.NoNode error if that file doesn't exist.

Reply:
This is a different situation: let's say SrvKeyspace has been rebuilt across all cells for ks1. Now, if we create ks2, it will already exist, and this code path will get skipped. But we actually need to rebuild to get the ks2 part of the SrvKeyspace.

Reply:
nvm. SrvKeyspace is the keyspace we're looking for.

Reply:
To clarify for anyone else following along:
- SrvKeyspace is stored at .../keyspaces/\<keyspace\>/SrvKeyspace
- We already detect if that file doesn't exist and call Rebuild
- Sometimes (for unknown reasons), the file exists but is empty, which leads to vtgate throwing a `no shard in keyspace` error.
- This PR is intended to detect that situation.

Reply:
At Slack, we once ran into a very bad incident due to this current logic of rebuilding the keyspace automatically. Since then, I'm nervous of this approach. I would advocate against adding more of those in the codebase. 

In my current view of the world, this kind of errors should fold into the category where an operator intervention is expected. If that unknown reason manifest, the operator should manually call RebuildKeyspaceGraph to fix the issue. 

I'm going to request changes to this PR since Sugu approved, so we can talk more about this :D